### DISCUSSION TITLE
Integrating cppwamp into a multithreaded app

### DISCUSSION THREAD

**Discussion by yupuquiz** [ecorm/cppwamp#122] [General]
Hello,

Just a question about integrating cppwap library into a multithreaded app.
publish an event is a thread safe? how could be the library use when it is integrated in a multi-thread app?

Thanks

**Comment by ecorm** [comment]
No, publishing an event is currently not thread-safe. Nor are any of `Session`'s (and `CoroSession`'s) public member functions. You need to serialize these calls coming from multiple threads by queuing them in the `boost::io_service` using `boost::io_service::post`:

```
Event event("myapp.foo").withArgs(42, otherStuff);
iosvc.post( [event]() {session->publish(event);} );
```

where `iosvc` is the same `io_service` used by the `Session` or `CoroSession`.

You might also be able use mutexes to serialize the `publish` calls, but I find the `io_service` approach to be less error-prone. Furthermore, mutex contention might degrade the performance of your worker threads.

Please note that if you have several `Session` / `CoroSession` instances, then they can be safely used in separate threads (I've never tested this, though).

I have an open issue #89 to consider making `Session` thread-safe, but I don't know when I'll have the time to look into this.

**Comment by yupuquiz** [comment]
ok, thanks for your explanation.

I guess to call a remote procedure is not thread-safe as well.

**Comment by ecorm** [comment]
Like I said, none of the public `Session` methods are thread-safe, including calling an RPC.

**Comment by ecorm** [comment]
You may find this [question](https://github.com/ecorm/cppwamp/issues/85), and this [Google Groups discussion](https://groups.google.com/forum/#!topic/cppwamp/9oUi_4sBRQs) to be informative.

**Comment by ecorm** [comment]
v0.10.0 now has thread-safe method overloads for `Session` taking the `ThreadSafe` tag type. These thread-safe overloads take care of serializing the operation via `Session::strand()`.

It's also now possible to pass the `boost::asio::thread_pool` context to `Session::create`. User handlers would then be `post`ed to that thread pool automatically.

It's also now possible to use `boost::asio::bind_executor` in a more fine-grained manner to make `Session` post a handler via a thread pool's executor.
